machine/src/chip.rs

```diff
@@ -22,31 +19,310 @@ pub trait Chip<M: Machine> {
         vec![]
     }
 
-    fn global_sends(&self, machine: &M) -> Vec<Interaction<M::F>> {
+    fn global_sends(&self, _machine: &M) -> Vec<Interaction<M::F>> {
         vec![]
     }
 
-    fn global_receives(&self, machine: &M) -> Vec<Interaction<M::F>> {
+    fn global_receives(&self, _machine: &M) -> Vec<Interaction<M::F>> {
         vec![]
     }
+
+    fn all_interactions(&self, machine: &M) -> Vec<(Interaction<M::F>, InteractionType)> {
+        let mut interactions: Vec<(Interaction<M::F>, InteractionType)> = vec![];
+        interactions.extend(
+            self.local_sends()
+                .into_iter()
+                .map(|i| (i, InteractionType::LocalSend)),
+        );
+        interactions.extend(
+            self.local_receives()
+                .into_iter()
+                .map(|i| (i, InteractionType::LocalReceive)),
+        );
+        interactions.extend(
+            self.global_sends(machine)
+                .into_iter()
+                .map(|i| (i, InteractionType::GlobalSend)),
+        );
+        interactions.extend(
+            self.global_receives(machine)
+                .into_iter()
+                .map(|i| (i, InteractionType::GlobalReceive)),
+        );
+        interactions
+    }
+
+    fn interaction_map(&self, machine: &M) -> BTreeMap<BusArgument, Vec<usize>> {
+        let mut map: BTreeMap<BusArgument, Vec<usize>> = BTreeMap::new();
+        for (n, (interaction, _)) in self.all_interactions(machine).iter().enumerate() {
+            map.entry(interaction.argument_index)
+                .or_insert_with(Vec::new)
+                .push(n);
+        }
+        map
+    }
+}
+
+pub trait ValidaAir<AB: PermutationAirBuilder, M: Machine> {
```

Comment:
I think we discussed a bit how if `ValidaAirBuilder` gives access to `Machine`, we could use the normal `Air` API, something like

```rust
pub trait ValidaAirBuilder<M: Machine>: PermutationAirBuilder {
    fn machine(&self) -> &M;
}

impl<M, AB> Air<AB> for SomeChip
where
    M: Machine + ...,
    AB: ValidaAirBuilder<M>,
{
    fn eval(&self, builder: &mut AB) {
        // normal constraints
        // bus interactions
    }
}
```

I think either approach would work fine though, WDYT?

Reply:
I'd like to avoid having custom subtraits if they aren't needed, so in favor of removing and going with this approach. The compiler complains that `M` is not constrained if `Machine` is a generic though (and I also want to avoid adding generic types to a chip struct), so I've added it for now as an associated type in `ValidaAirBuilder`.